api/workload/x509_client.go

```diff
@@ -42,40 +56,164 @@ func NewX509Client(c *X509ClientConfig) *x509Client {
+	current *workload.X509SVIDResponse
 }
 
 func (x *x509Client) Start() error {
-	errChan := make(chan error, 1)
-	go func() { errChan <- x.stream.listen() }()
-	defer x.stream.stop()
-
-	select {
-	case <-x.stopChan:
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// make sure the client hasn't already been started
+	x.mu.Lock()
+	if x.cancel != nil {
+		x.mu.Unlock()
+		return errors.New("already started")
+	}
+	x.cancel = cancel
+	x.mu.Unlock()
+
+	// allow for another start after this function returns
+	defer func() {
+		x.mu.Lock()
+		x.cancel = nil
+		x.mu.Unlock()
+	}()
+
+	// set up a channel to receive the updates, update the current
+	// update, and push it down the update channel. the channel will
+	// be closed when this function returns to terminate this goroutine.
+	out := make(chan *workload.X509SVIDResponse)
+	defer close(out)
+	go func() {
+		for upd := range out {
+			// set the current update
+			x.mu.Lock()
+			x.current = upd
+			x.mu.Unlock()
+
+			// throw away an existing unread update
+			select {
+			case <-x.updateChan:
+			default:
+			}
+
+			// push the update down the channel
+			select {
+			case x.updateChan <- upd:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	err := streamX509SVID(ctx, x.c, out)
+	switch err {
+	case nil, context.Canceled:
 		return nil
-	case err := <-errChan:
+	default:
 		return err
 	}
 }
 
 func (x *x509Client) Stop() {
-	close(x.stopChan)
+	x.mu.RLock()
+	if x.cancel != nil {
+		x.cancel()
+	}
+	x.mu.RUnlock()
 }
 
 func (x *x509Client) CurrentSVID() (*workload.X509SVIDResponse, error) {
-	return x.stream.current()
+	x.mu.RLock()
+	defer x.mu.RUnlock()
+	if x.current == nil {
+		return nil, errors.New("no SVID received yet")
+	}
+	return x.current, nil
 }
 
 func (x *x509Client) UpdateChan() <-chan *workload.X509SVIDResponse {
-	return x.stream.updateChan()
+	return x.updateChan
+}
+
+func streamX509SVID(ctx context.Context, config *X509ClientConfig, out chan<- *workload.X509SVIDResponse) error {
+	ctx = metadata.NewOutgoingContext(ctx, metadata.Pairs("workload.spiffe.io", "true"))
+
+	conn, err := dial.Dial(ctx, config.Addr)
+	if err != nil {
+		// dial shouldn't fail unless there is a misconfiguration
+		return err
+	}
+	defer conn.Close()
+
+	// TODO: jitter?
+	const backoffMin = time.Second
+	backoff := backoffMin
+
+	handleErr := func(ctx context.Context, op string, err error) error {
+		switch status.Code(err) {
+		case codes.DeadlineExceeded, codes.Canceled:
+			return ctx.Err()
+		case codes.InvalidArgument:
+			return err
+		}
+		if config.FailOnError {
+			config.log("%s failed with %v; aborting", op, err)
+			return err
+		}
+
+		config.log("%s failed with %v; retrying in %s", op, err, backoff)
+		timer := time.NewTimer(backoff)
+		defer timer.Stop()
+		select {
+		case <-timer.C:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		backoff = time.Duration(float64(backoff) * 1.5)
+		if backoff > config.Timeout {
+			backoff = config.Timeout
```

Comment:
IIRC the intention of the timeout was to limit how long we back off for e.g. if we have backed off for so long that our retry timer exceeds the timeout, then bail.... but, maybe the code I wrote didn't actually do that :)

Reply:
From what I could tell it didn't end up doing that (at least in the circumstances i tested). What I assumed the intent was is that FailOnError would cause it to bail after Timeout was exceeded, otherwise timeout would become the max backoff period (to control how fast it would respond if things turned green again). I'm happy to change the semantics to whatever we want, though.

Reply:
FailOnError was meant to not apply backoff at all

I'm up for whatever you think is the Right Thing... I'm not totally sure what the best option is, frankly